adapters/intertech/intertech.go

```diff
@@ -0,0 +1,326 @@
+func splitRequestDataByImp(request *openrtb2.BidRequest, imp openrtb2.Imp) openrtb2.BidRequest {
+	requestCopy := *request
+	requestCopy.Imp = []openrtb2.Imp{imp}
+
+	return requestCopy
+}
+
+func getPlacementID(imp openrtb2.Imp) (*intertechPlacementID, error) {
+	var ext adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &ext); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("imp %s: unable to unmarshal ext", imp.ID),
+		}
+	}
+
+	var intertechExt openrtb_ext.ExtImpIntertech
+	if err := json.Unmarshal(ext.Bidder, &intertechExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("imp %s: unable to unmarshal ext.bidder: %v", imp.ID, err),
+		}
+	}
+
+	placementID := mapExtToPlacementID(intertechExt)
+	return placementID, nil
+}
+
+func mapExtToPlacementID(intertechExt openrtb_ext.ExtImpIntertech) *intertechPlacementID {
+	var placementID intertechPlacementID
+
+	if len(intertechExt.PlacementID) == 0 {
+		placementID.ImpID = strconv.Itoa(int(intertechExt.ImpID))
+		placementID.PageID = strconv.Itoa(int(intertechExt.PageID))
+		return &placementID
+	}
+
+	idParts := strings.Split(intertechExt.PlacementID, "-")
+	placementID.PageID = idParts[0]
+	placementID.ImpID = idParts[1]
+
+	return &placementID
+}
+
+func modifyImp(imp *openrtb2.Imp) error {
+	if imp.Banner != nil {
+		banner, err := modifyBanner(*imp.Banner)
+		if banner != nil {
+			imp.Banner = banner
+		}
+		return err
+	}
+
+	if imp.Native != nil {
+		return nil
+	}
+
+	return &errortypes.BadInput{
+		Message: fmt.Sprintf("Unsupported format. Intertech only supports banner and native types. Ignoring imp id #%s", imp.ID),
+	}
+}
+
+func modifyBanner(banner openrtb2.Banner) (*openrtb2.Banner, error) {
+	format := banner.Format
+
+	hasRootSize := banner.W != nil && banner.H != nil && *banner.W > 0 && *banner.H > 0
+	if !hasRootSize && len(format) == 0 {
+		w := 0
+		h := 0
+		if banner.W != nil {
+			w = int(*banner.W)
+		}
+		if banner.H != nil {
+			h = int(*banner.H)
+		}
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("Invalid sizes provided for Banner %dx%d", w, h),
+		}
+	}
+
+	if !hasRootSize {
+		firstFormat := format[0]
+		banner.H = &firstFormat.H
+		banner.W = &firstFormat.W
+	}
+
+	return &banner, nil
+}
+
+// resolveUrl "un-templates" the endpoint by replacing macroses and adding the required query parameters
+func (a *adapter) resolveUrl(placementID intertechPlacementID, referer string, currency string) (string, error) {
+	params := macros.EndpointTemplateParams{PageID: placementID.PageID, ImpID: placementID.ImpID}
+
+	endpointStr, err := macros.ResolveMacros(a.endpoint, params)
+	if err != nil {
+		return "", err
+	}
+
+	parsedUrl, err := url.Parse(endpointStr)
+	if err != nil {
+		return "", err
+	}
+
+	addNonEmptyQueryParams(parsedUrl, map[string]string{
+		refererQueryKey:  referer,
+		currencyQueryKey: currency,
+		impIdQueryKey:    placementID.ImpID,
+	})
+
+	return parsedUrl.String(), nil
```

Comment:
Curious, what is the purpose of parsing the url and converting it back to a string?

Reply:
Thanks for pointing this out! I believe url.Parse is used here primarily to validate the generated URL and to simplify adding query parameters with addNonEmptyQueryParams. However, if this isn't necessary, I can refactor it to work directly with strings. Let me know if that would be preferable.

Reply:
Gotcha. I think it's a good idea to keep it for adding query parameters.